research/utils/logging.py

```diff
@@ -1,32 +1,46 @@
+from typing import Callable, List
 from ignite.engine import Engine
+from torch.utils.data import DataLoader
 
 
-# FIXME: Use logging module, not print statements
-def log_training_loss(engine: Engine) -> None:
-    print("Epoch[{}] Loss: {:.2f}".format(engine.state.epoch, engine.state.output[0]))
+def _lf(log_fn: Callable[[Engine, List[str]], None], fields: List[str]):
+    """Returns a log function lambda. Useful for code clarity"""
+    return lambda _: log_fn(_, fields=fields)
 
 
-# FIXME: Use logging module, not print statements
-def log_training_results(engine: Engine) -> None:
-    evaluator.run(train_loader)
-    metrics = evaluator.state.metrics
-    print(
-        "Training Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}".format(
-            engine.state.epoch, metrics["accuracy"], metrics["nll"]
-        )
-    )
+def _lf_val(
+    log_fn: Callable[[Engine, List[str]], None],
+    engine: Engine,
+    loader: DataLoader,
+    fields: List[str],
+):
+    """Returns a log function lambda. Useful for code clarity"""
+    return lambda _: log_fn(_, engine, loader, fields=fields)
 
 
-# FIXME: Use logging module, not print statements
-def log_validation_results(engine: Engine) -> None:
-    evaluator.run(val_loader)
-    metrics = evaluator.state.metrics
-    print(
-        "Validation Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}".format(
-            engine.state.epoch, metrics["accuracy"], metrics["nll"]
-        )
+def log_engine_output_stdout(engine: Engine, fields: List[str]) -> None:
```

Comment:
I see what you are trying to do i think. The logger is essential some human readable state representation of the engine. But instead of having this method of extracting state shouldn't the engine itself output the state representation? I say this partly because what you can log right now is only the state at 1 single iteration. But there are time it is good to see some rolling average of the loss. Essentially if we want to see how the state has changed over the past N iterations to see a smoothed less variation graph.

Reply:
From the Ignite standpoint, moving average is implemented as a `Metric`. We can do what you describe - a moving average over the state at the iteration or epoch level. 

https://pytorch.org/ignite/metrics.html#ignite.metrics.RunningAverage

You raise a better point. It should be intuitive to whoever clone this code _how_ to do that.

Reply:
So if you wanted a running average and log that then you would have to put it in the _metric_transform?

Reply:
Depends, is the running average taken over the training set, or the validation set? As currently written, I have no metrics on the training set, only the validation set. We can set metrics on any engine (training/validation/etc.), and we can set a moving average on those metrics.

https://pytorch.org/ignite/_modules/ignite/metrics/running_average.html

Reply:
okay then this current method doesn't seem too bad since custom metrics can be defined. Maybe it could be done better but it isn't bad right now